File: src/hollywood/flows.py
from __future__ import annotations

from collections import defaultdict

from prefect import flow, task

from .adapters import get_adapter
from .archive import archive_payload
from .config import HollywoodSettings
from .models import IngestOptions, RunStatus, RunSummary, SourceDefinition
from .registry import get_source, list_sources
from .storage import HollywoodStorage


@task
def fetch_payloads_task(
    source: SourceDefinition, settings: HollywoodSettings, options: IngestOptions
):
    adapter = get_adapter(source)
    return adapter.fetch_raw_payloads(settings, options)


@task
def archive_payloads_task(source: SourceDefinition, settings: HollywoodSettings, payloads):
    return [archive_payload(settings, source, payload) for payload in payloads]


@task
def normalize_payloads_task(
    source: SourceDefinition,
    settings: HollywoodSettings,
    storage: HollywoodStorage,
    run_id: str,
    raw_records: list[dict[str, object]],
):
    adapter = get_adapter(source)
    return adapter.normalize_raw_records(settings, storage, run_id, raw_records)


@flow(name="hollywood-ingest-source")
def ingest_source_flow(
    source_id: str, settings: HollywoodSettings, options: IngestOptions
) -> RunSummary:
    source = get_source(source_id)
    storage = HollywoodStorage(settings.resolved_db_path)
    storage.initialize()
    run_id = storage.start_run(source, options.model_dump_json())
    try:
        payloads = fetch_payloads_task(source, settings, options)
        archived_payloads = archive_payloads_task(source, settings, payloads)
        storage.insert_raw_records(run_id, archived_payloads)
        raw_records = storage.load_raw_records(run_id=run_id)
        bundle = normalize_payloads_task(source, settings, storage, run_id, raw_records)
        storage.apply_bundle(bundle)
        summary = RunSummary(
            run_id=run_id,
            source_id=source.source_id,
            status=RunStatus.SUCCEEDED,
            raw_records=len(archived_payloads),
            normalized=bundle.counts(),
        )
        storage.finish_run(run_id, RunStatus.SUCCEEDED, summary.model_dump(mode="json"))
        return summary
    except Exception as exc:
        failure_summary = {"source_id": source.source_id, "error": str(exc)}
        storage.finish_run(run_id, RunStatus.FAILED, failure_summary, error_text=str(exc))
        raise


@flow(name="hollywood-ingest-group")
def ingest_group_flow(
    group_name: str, settings: HollywoodSettings, options: IngestOptions
) -> list[RunSummary]:
    summaries: list[RunSummary] = []
    for source in list_sources(group=group_name):
        summaries.append(ingest_source_flow(source.source_id, settings, options))
    return summaries


@flow(name="hollywood-normalize")
def normalize_flow(source_id: str | None, settings: HollywoodSettings) -> dict[str, int]:
    storage = HollywoodStorage(settings.resolved_db_path)
    storage.initialize()
    grouped_records: dict[str, list[dict[str, object]]] = defaultdict(list)
    for record in storage.load_raw_records(source_id=source_id):
        grouped_records[str(record["source_id"])].append(record)

    combined_counts = defaultdict(int)
    for grouped_source_id, records in grouped_records.items():
        source = get_source(grouped_source_id)
        adapter = get_adapter(source)
        run_id = f"normalize::{grouped_source_id}"
        bundle = adapter.normalize_raw_records(settings, storage, run_id, records)
        storage.apply_bundle(bundle)
        for key, value in bundle.counts().items():
            combined_counts[key] += value
    return dict(combined_counts)


def export_flow(
    settings: HollywoodSettings, file_format: str, table: str | None = None
) -> list[str]:
    storage = HollywoodStorage(settings.resolved_db_path)
    storage.initialize()
    if table:
        return [str(storage.export_table(table, settings.parquet_dir, file_format))]
    return [str(path) for path in storage.export_all(settings.parquet_dir, file_format)]

File: api/src/ingest/flows.ts
import type { RawRecordRow } from '../db/repositories/RawRecordRepository.js';
import { IngestService } from '../db/services/IngestService.js';
import type { Adapter } from './adapters/base.js';
import { archivePayload } from './archive.js';
import { bundleCounts } from './models.js';
import type { DoctorCheck, IngestOptions, RunSummary } from './models.js';
import { getSource, listSources } from './registry.js';

const ADAPTERS = new Map<string, Adapter>();

const ingestService = new IngestService();

export function registerAdapter(sourceId: string, adapter: Adapter): void {
  ADAPTERS.set(sourceId, adapter);
}

function getAdapter(sourceId: string): Adapter {
  const adapter = ADAPTERS.get(sourceId);
  if (!adapter) {
    throw new Error(
      `No TypeScript adapter registered yet for source '${sourceId}' (ports land in a later phase)`,
    );
  }
  return adapter;
}

export async function normalizeFlow(sourceId?: string): Promise<Record<string, number>> {
  const rawRecords = ingestService.loadRawRecords(sourceId ? { sourceId } : {});
  const grouped = new Map<string, RawRecordRow[]>();
  for (const record of rawRecords) {
    const key = record.sourceId;
    if (!grouped.has(key)) grouped.set(key, []);
    grouped.get(key)!.push(record);
  }

  const combinedCounts: Record<string, number> = {};
  for (const [groupedSourceId, records] of grouped) {
    getSource(groupedSourceId);
    const adapter = getAdapter(groupedSourceId);
    const runId = ingestService.startRunRaw('normalize', { source_id: groupedSourceId });
    const bundle = await adapter.normalizeRawRecords(runId, records);
    ingestService.applyBundle(bundle);
    const counts = bundleCounts(bundle);
    ingestService.finishRun(runId, 'succeeded', counts);
    for (const [key, value] of Object.entries(counts)) {
      combinedCounts[key] = (combinedCounts[key] ?? 0) + value;
    }
  }
  return combinedCounts;
}

export function sourceDoctorChecks(): DoctorCheck[] {
  const checks: DoctorCheck[] = [];
  for (const source of listSources()) {
    const adapter = ADAPTERS.get(source.sourceId);
    if (adapter?.doctorChecks) {
      checks.push(...adapter.doctorChecks());
    } else if (adapter) {
      checks.push({
        name: `${source.sourceId}:config`,
        ok: true,
        detail: `Configured fetch strategy: ${source.fetchStrategy}`,
      });
    } else {
      checks.push({
        name: `${source.sourceId}:config`,
        ok: true,
        detail: `Configured fetch strategy: ${source.fetchStrategy} (adapter not yet ported)`,
      });
    }
  }
  return checks;
}

export async function runIngestSource(
  sourceId: string,
  options: IngestOptions,
): Promise<RunSummary> {
  const source = getSource(sourceId);
  const adapter = getAdapter(sourceId);
  const runId = ingestService.startRun(source.sourceId, JSON.stringify(options));
  try {
    const payloads = await adapter.fetchRawPayloads(options);
    const archivedPayloads = payloads.map((p) => archivePayload(source, p));
    ingestService.insertRawRecords(runId, archivedPayloads);
    const rawRecords = ingestService.loadRawRecords({ runId });
    const bundle = await adapter.normalizeRawRecords(runId, rawRecords);
    ingestService.applyBundle(bundle);
    const summary: RunSummary = {
      runId,
      sourceId: source.sourceId,
      status: 'succeeded',
      rawRecords: archivedPayloads.length,
      normalized: bundleCounts(bundle),
    };
    ingestService.finishRun(runId, 'succeeded', summary);
    return summary;
  } catch (exc) {
    const error = exc instanceof Error ? exc.message : String(exc);
    ingestService.finishRun(runId, 'failed', { source_id: source.sourceId, error }, error);
    throw exc;
  }
}

export async function runIngestGroup(
  groupName: string,
  options: IngestOptions,
): Promise<RunSummary[]> {
  const summaries: RunSummary[] = [];
  for (const source of listSources(groupName)) {
    summaries.push(await runIngestSource(source.sourceId, options));
  }
  return summaries;
}
